### DISCUSSION TITLE
Are there any ways to use encryption for ldap_bindpassword in config.json?

### DISCUSSION THREAD

**Discussion by Ye-Min-Tun** [semaphoreui/semaphore#1728] [Q&A] [answered]
I feel like setting plain text in ldap_bindpassword in config.json is highly vulnerable for security. Are you guys using it as plain text? Can I use any encryption method for it?

**Comment by tboerger** [comment]
That's pretty normal for any kind of application. Where should the encryption key get stored? Instead of storing it within the config.json I think it would be great to source all secrets from separate files, that's something I have started to implement within https://github.com/ansible-semaphore/semaphore/pull/1517 for the OIDC integration.

**Reply by Ye-Min-Tun** [reply]
Do you have plan to implement it also for LDAP?

**Reply by tboerger** [reply]
First @fiftin got to accept my contribution for OIDC before I would introduce it somewhere else. In the meantime I would suggest the mentioned environment variable.

**Comment by tboerger** [comment]
You could also store the LDAP password within an environment variable instead of the configuration file, this should be `SEMAPHORE_LDAP_BIND_PASSWORD` defined at https://github.com/ansible-semaphore/semaphore/blob/34485b7b8e825a1a1f0e671e3d2a876c7d1c1796/util/config.go#L153-L160. The only part that can't get set via environment variables is the LDAP mapping.

**Reply by Ye-Min-Tun** [reply]
@tboerger Thanks a lot for answering.
This would also viewable by all Semaphore project users, right? I don't want to let that password see by the users.

**Reply by tboerger** [reply]
Users in general? Not really since the don't have direct access to environment variables or configuration files. If you are afraid about malicious playbooks reading secrets from the Semaphore instance you could start dedicated runner instances which won't have access to any Semaphore configuration. Take a look at https://docs.semui.co/administration-guide/runners to get an overview about runners.

**Reply by canlot** [reply]
No, unless your users have access to the server where all files are. But it would be a bad practice and not recommended.

**Reply by Ye-Min-Tun** [reply]
Where should I set this env:"SEMAPHORE_LDAP_BIND_PASSWORD" variable? I thought that was the one we can set in UI > Environment. That's why I said all the project users can see this ENV var.  Actually, I might have a wrong logic. How can the LDAP users login without that password?

but even with the runners, we need to key in the ldap_bindpassword as plain text in our semaphore server, right? My problem is that the server team doesn't allow to key in the plain text for security concern. @tboerger

**Reply by Ye-Min-Tun** [reply]
@canlot The user here means the ones who can access the server, not the semaphore user. Sorry for my vague description.

**Reply by tboerger** [reply]
But than you are not able to install any application which includes some kind of secret... Show me any application that can store LDAP search bindings that are stored encrypted.

It is absolutely normal to provide secrets in form of an envrionment variable or configuration file to the application process. Generally these secrets are only visible to administrators that got access to the server, not to users of Semaphore.

The runner integration helps against malicious playbooks executed by Semaphore as this runner should be deployed separately from the Semaphore server, thus it won't give any access to the mentioned secrets.

Maybe you could explain more in detail the problem of the server team. The required secret ist used for the bind user to allow searching the directory, that's totally normal for applications connected to an LDAP.

**Reply by Ye-Min-Tun** [reply]
Thanks for explanation. I understand what you mentioned.
For security reason, we cannot keep credentials in config file.
Is it possible to set that password by using the **key store** mechanism or else in UI? 

I also have a confusion. This LDAP integration is for Semaphore Login, not for the remote hosts, right? @tboerger

**Reply by tboerger** [reply]
If not within config file than use environment variables until it could be stored in dedicated secret files. The key store is used for the Ansible runtimes, not for the Semaphore server itself.

That's right, the password is used for the LDAP-based authentication, it is not used for any host authentication at all.

**Reply by Ye-Min-Tun** [reply]
Thanks again for your kind reply.

I thought environment variable means the "Environment" mechanism in Semaphore UI. 
Do you mean it the server's environment variable?

**Reply by tboerger** [reply]
Exactly, I wasn't referring to Semaphore environments, I was referring to environment variables added to the server process at start.

**Reply by canlot** [reply, marked as answer]
Generally no one except the admins should access server, even if some users have access to server they shouldn't have root access. If semaphore running under certain account or under root account and the permission are set right there is no problem.

Even if the credentials from the bind account have been leaked this is usually not a problem because the user should be a dumb ldap user just created for the purpose of being bind to ldap and not used for different applications. It should have permissions to list all users and groups from ldap server and nothing else. 

What you should consider for better security is, to use ldap over ssl using port 636 and not plain ldap. If you use plain ldap the credentials from all users who authenticate will be visible to anybody if they capture network traffic from the server.
